src/kv/RocksDBStore.cc

```diff
@@ -2663,3 +2663,377 @@ RocksDBStore::WholeSpaceIterator RocksDBStore::get_default_cf_iterator()
+    }
+  }
+  return 0;
+}
+
+int RocksDBStore::reshard(const std::string& new_sharding)
+{
+  rocksdb::Status status;
+  int r;
+  std::vector<std::string> to_process_columns;
+  std::vector<rocksdb::ColumnFamilyHandle*> to_process_handles;
+
+  size_t bytes_in_batch = 0;
+  size_t keys_in_batch = 0;
+  size_t bytes_per_iterator = 0;
+  size_t keys_per_iterator = 0;
+
+  rocksdb::WriteBatch* bat = nullptr;
+
+  auto flush_batch = [&]() {
+    dout(10) << "flushing batch" << dendl;
+    rocksdb::WriteOptions woptions;
+    woptions.sync = true;
+    rocksdb::Status s = db->Write(woptions, bat);
+    ceph_assert(s.ok());
+    dout(25) << "processed " << keys_in_batch << " keys, for " << bytes_in_batch << " bytes" << dendl;
+    bytes_in_batch = 0;
+    keys_in_batch = 0;
+    delete bat;
+    bat = new rocksdb::WriteBatch();
+    ceph_assert(bat);
+  };
+
+  auto process_column = [&](rocksdb::ColumnFamilyHandle* handle,
+			    const std::string& fixed_prefix)
+  {
+    dout(10) << " column=" << (void*)handle << " prefix=" << fixed_prefix << dendl;
+    rocksdb::Iterator* it;
+    it = db->NewIterator(rocksdb::ReadOptions(), handle);
+    ceph_assert(it);
+    bat = new rocksdb::WriteBatch();
+    ceph_assert(bat);
+
+    for (it->SeekToFirst(); it->Valid(); it->Next()) {
+      rocksdb::Slice raw_key = it->key();
+      dout(30) << "key=" << pretty_binary_string(raw_key.ToString()) << dendl;
+      //check if need to refresh iterator
+      if (bytes_per_iterator > 10000000 ||
+	  keys_per_iterator > 10000) {
+	dout(10) << "refreshing iterator" << dendl;
+	bytes_per_iterator = 0;
+	keys_per_iterator = 0;
+	std::string raw_key_str = raw_key.ToString();
+	delete it;
+	it = db->NewIterator(rocksdb::ReadOptions(), handle);
+	ceph_assert(it);
+	it->Seek(raw_key_str);
+	ceph_assert(it->Valid());
+      }
+      rocksdb::Slice value = it->value();
+      std::string prefix, key;
+      if (fixed_prefix.size() == 0) {
+	split_key(raw_key, &prefix, &key);
+      } else {
+	prefix = fixed_prefix;
+	key = raw_key.ToString();
+      }
+      std::string new_raw_key;
+      rocksdb::ColumnFamilyHandle* new_handle = get_cf_handle(prefix, key);
+      if (new_handle == nullptr) {
+	new_handle = default_cf;
+      }
+      if (handle == new_handle) {
+	continue;
+      }
+      if (new_handle == default_cf) {
+	new_raw_key = combine_strings(prefix, key);
+      } else {
+	new_raw_key = key;
+      }
+      bat->SingleDelete(handle, raw_key);
+      bat->Put(new_handle, new_raw_key, value);
+      dout(25) << "moving " << (void*)handle << "/" << pretty_binary_string(raw_key.ToString()) <<
+	" to " << (void*)new_handle << "/" << pretty_binary_string(new_raw_key) <<
+	" size " << value.size() << dendl;
+
+      bytes_in_batch += new_raw_key.size() * 2 + value.size();
+      keys_in_batch ++;
+      bytes_per_iterator += new_raw_key.size() * 2 + value.size();
+      keys_per_iterator ++;
+
+      //check if need to write batch
+      if (bytes_in_batch > 1000000 ||
```

Comment:
should we make these dev options? could see potentially wanting to do it in larger batches for speed, or smaller to throttle things.